File: api/waitlist.js
// api/waitlist.js — Phase 2 backend for waitlistService.js
// Tables: member_waitlist, cancellation_risk, demand_heatmap, bookings, members
// Return shapes IDENTICAL to waitlistService.js

import { sql } from '@vercel/postgres';
import { withAuth, getClubId } from './lib/withAuth.js';

const normalizeHealthScore = (value) => {
  const numeric = Number(value);
  if (!Number.isFinite(numeric)) return 0;
  const normalized = numeric <= 1 ? numeric * 100 : numeric;
  return Math.round(normalized);
};

const deriveRiskLevel = (score) => {
  if (score >= 70) return 'Healthy';
  if (score >= 50) return 'Watch';
  if (score >= 30) return 'At Risk';
  return 'Critical';
};

export default withAuth(async function handler(req, res) {
  const clubId = getClubId(req);
  try {
    const [queue, predictions, heatmap, revenueAttr] = await Promise.all([

      sql`
        SELECT
          mw.waitlist_id,
          mw.member_id::text AS member_id,
          COALESCE(NULLIF(TRIM(m.first_name || ' ' || m.last_name), ''), 'Member ' || RIGHT(mw.member_id::text, 3)) AS member_name,
          m.archetype,
          m.annual_dues AS member_value_annual,
          mw.requested_slot, mw.days_waiting, mw.alternatives_accepted,
          mw.retention_priority, mw.dining_incentive_attached,
          mw.notified_at, mw.filled_at,
          w.engagement_score AS health_score,
          CASE WHEN w.engagement_score >= 70 THEN 'Healthy'
               WHEN w.engagement_score >= 50 THEN 'Watch'
               WHEN w.engagement_score >= 30 THEN 'At Risk'
               ELSE 'Critical' END AS risk_level,
          (SELECT MAX(b.booking_date) FROM bookings b
           JOIN booking_players bp ON b.booking_id = bp.booking_id AND bp.club_id = ${clubId}
           WHERE bp.member_id::text = mw.member_id::text AND b.status = 'completed' AND b.club_id = ${clubId}) AS last_round,
          (SELECT COUNT(*) FILTER (WHERE pc.post_round_dining = 1)::float /
                  NULLIF(COUNT(*), 0)
           FROM pos_checks pc WHERE pc.member_id::text = mw.member_id::text AND pc.club_id = ${clubId}) AS dining_history_rate
        FROM member_waitlist mw
        JOIN members m ON mw.member_id::text = m.member_id::text AND m.club_id = ${clubId}
        JOIN member_engagement_weekly w ON m.member_id::text = w.member_id::text AND w.club_id = ${clubId}
          AND w.week_number = (SELECT MAX(week_number) FROM member_engagement_weekly WHERE club_id = ${clubId})
        WHERE mw.club_id = ${clubId}
        ORDER BY CASE mw.retention_priority WHEN 'HIGH' THEN 0 ELSE 1 END, w.engagement_score ASC`,

      sql`
        SELECT cr.risk_id, cr.booking_id, cr.member_id::text AS member_id,
               COALESCE(NULLIF(TRIM(m.first_name || ' ' || m.last_name), ''), 'Member ' || RIGHT(cr.member_id::text, 3)) AS member_name,
               m.archetype,
               b.tee_time, b.booking_date,
               cr.cancel_probability, cr.drivers, cr.recommended_action,
               cr.estimated_revenue_lost, cr.action_taken, cr.outcome
        FROM cancellation_risk cr
        JOIN bookings b ON cr.booking_id = b.booking_id AND b.club_id = ${clubId}
        JOIN members m ON cr.member_id::text = m.member_id::text AND m.club_id = ${clubId}
        WHERE cr.club_id = ${clubId} AND b.booking_date::date >= '2026-01-01'::date AND b.status = 'confirmed'
        ORDER BY cr.cancel_probability DESC LIMIT 20`,

      sql`
        SELECT dh.heatmap_id, c.name AS course, dh.day_of_week, dh.time_block,
               dh.fill_rate, dh.unmet_rounds, dh.demand_level
        FROM demand_heatmap dh
        JOIN courses c ON dh.course_id = c.course_id AND c.club_id = ${clubId}
        WHERE dh.club_id = ${clubId}
        ORDER BY c.name,
          ARRAY_POSITION(ARRAY['Mon','Tue','Wed','Thu','Fri','Sat','Sun'], dh.day_of_week),
          dh.time_block`,

      sql`
        SELECT
          ROUND(AVG(vs.total_spend) FILTER (WHERE vs.anchor_type = 'golf')::numeric, 2) AS avg_golf_spend,
          ROUND(AVG(vs.total_spend) FILTER (WHERE vs.anchor_type = 'golf' AND vs.touchpoints >= 2)::numeric, 2) AS avg_multi_touch_spend
        FROM visit_sessions vs WHERE vs.club_id = ${clubId}`,
    ]);

    const queueRows = queue.rows.map(w => {
      const healthScore = normalizeHealthScore(w.health_score);
      const alternatives = Array.isArray(w.alternatives_accepted)
        ? w.alternatives_accepted
        : JSON.parse(w.alternatives_accepted ?? '[]');
      return {
        waitlistId:           w.waitlist_id,
        memberId:             w.member_id,
        memberName:           w.member_name,
        archetype:            w.archetype,
        memberValueAnnual:    Number(w.member_value_annual) || 0,
        requestedSlot:        w.requested_slot,
        daysWaiting:          Number(w.days_waiting),
        alternatesAccepted:   alternatives,
        retentionPriority:    w.retention_priority,
        diningIncentiveAttached: Number(w.dining_incentive_attached) === 1 || w.dining_incentive_attached === true,
        notifiedAt:           w.notified_at,
        filledAt:             w.filled_at,
        healthScore,
        riskLevel:            deriveRiskLevel(healthScore),
        lastRound:            w.last_round,
        diningHistory:        Number(w.dining_history_rate) > 0.35 ? 'High converter'
                            : Number(w.dining_history_rate) > 0.20 ? 'Moderate' : 'Low converter',
      };
    });

    const predRows = predictions.rows.map(p => ({
      riskId:               p.risk_id,
      bookingId:            p.booking_id,
      memberId:             p.member_id,
      memberName:           p.member_name,
      archetype:            p.archetype,
      teeTime:              p.tee_time,
      bookingDate:          p.booking_date,
      cancelProbability:    Number(p.cancel_probability),
      drivers:              JSON.parse(p.drivers ?? '[]'),
      recommendedAction:    p.recommended_action,
      estimatedRevenueLost: Number(p.estimated_revenue_lost),
      actionTaken:          p.action_taken,
      outcome:              p.outcome,
    }));

    const highRisk = predRows.filter(p => p.cancelProbability >= 0.60);
    const atRiskQueue = queueRows.filter((entry) => ['At Risk', 'Critical'].includes(entry.riskLevel));

    res.status(200).json({
      queue: queueRows,
      summary: {
        total: queueRows.length,
        highPriority: queueRows.filter(w => w.retentionPriority === 'HIGH').length,
        atRisk: atRiskQueue.length,
        critical: queueRows.filter((w) => w.riskLevel === 'Critical').length,
        avgDaysWaiting: queueRows.length
          ? Math.round(queueRows.reduce((s, w) => s + w.daysWaiting, 0) / queueRows.length) : 0,
        riskScoredToday: queueRows.filter((w) => Number.isFinite(w.healthScore) && w.healthScore > 0).length,
        atRiskDuesExposed: Math.round(atRiskQueue.reduce((sum, row) => sum + (row.memberValueAnnual || 0), 0)),
      },
      cancellationPredictions: predRows,
      cancellationSummary: {
        total: predRows.length, highRisk: highRisk.length,
        totalRevAtRisk: Math.round(highRisk.reduce((s, p) => s + p.estimatedRevenueLost, 0)),
        topDriver: 'Wind advisory + low-engagement members',
      },
      demandHeatmap: heatmap.rows.map(h => ({
        heatmapId: h.heatmap_id, course: h.course, dayOfWeek: h.day_of_week,
        timeBlock: h.time_block, fillRate: Number(h.fill_rate),
        unmetRounds: Number(h.unmet_rounds), demandLevel: h.demand_level,
      })),
      revenueAttribution: {
        reactive:          Number(revenueAttr.rows[0]?.avg_golf_spend ?? 187),
        retentionPriority: Number(revenueAttr.rows[0]?.avg_multi_touch_spend ?? 312),
        upliftPct: Math.round(
          ((Number(revenueAttr.rows[0]?.avg_multi_touch_spend ?? 312) /
            Math.max(1, Number(revenueAttr.rows[0]?.avg_golf_spend ?? 187))) - 1) * 100
        ),
      },
    });
  } catch (err) {
    console.error('/api/waitlist error:', err);
    res.status(500).json({ error: err.message });
  }
}, { allowDemo: true });

File: seed/generators/waitlist.py
"""
seed/generators/waitlist.py
Phase 4 — Waitlist & Demand domain (Sprint Addition)
  member_waitlist (~10 entries), cancellation_risk (~35), demand_heatmap (~56)

The 5 at-risk resignation members are prioritized in the queue.
Cancellation risk scored for all upcoming confirmed bookings.
"""
import random
import json
from datetime import date, datetime, timedelta

PRIORITY_THRESHOLD = 50   # health score below this = HIGH priority
WIND_THRESHOLD_MPH = 15   # wind above this triggers weather cancel signal

# At-risk resignation members anchor the waitlist
AT_RISK_MEMBERS = [
    {
        'member_id': 'mbr_059', 'name': 'Anne Jordan',
        'archetype': 'Weekend Warrior', 'health_score': 38,
        'risk_level': 'At Risk', 'last_round': '2025-12-28',
        'dining_history': 'Low converter',
    },
    {
        'member_id': 'mbr_071', 'name': 'Kevin Hurst',
        'archetype': 'Declining', 'health_score': 22,
        'risk_level': 'Critical', 'last_round': '2025-11-10',
        'dining_history': 'Stopped dining',
    },
    {
        'member_id': 'mbr_072', 'name': 'Steven Park',
        'archetype': 'Declining', 'health_score': 31,
        'risk_level': 'At Risk', 'last_round': '2025-10-30',
        'dining_history': 'Hits F&B minimum only',
    },
    {
        'member_id': 'mbr_038', 'name': 'James Whitfield',
        'archetype': 'Balanced Active', 'health_score': 44,
        'risk_level': 'Watch', 'last_round': '2026-01-14',
        'dining_history': 'Regular diner, complaint pending',
    },
    {
        'member_id': 'mbr_089', 'name': 'Linda Leonard',
        'archetype': 'Ghost', 'health_score': 8,
        'risk_level': 'Critical', 'last_round': '2025-09-15',
        'dining_history': 'Rare',
    },
]

# Additional healthy members for contrast (indices into member pool)
HEALTHY_CONTRAST_ARCHETYPES = [
    ('Die-Hard Golfer', 82, 'Healthy'),
    ('Balanced Active', 74, 'Healthy'),
    ('Social Butterfly', 78, 'Healthy'),
    ('New Member', 65, 'Watch'),
    ('Snowbird', 71, 'Healthy'),
]

REQUESTED_SLOTS = ['07:00', '07:08', '07:16', '07:24', '08:00', '08:32', '09:00']
ALT_SLOTS = [
    ['07:08', '07:16', '08:00'],
    ['07:00', '07:24', '08:08'],
    ['07:16', '08:00', '08:32'],
    ['07:00', '07:08', '09:00'],
]

CANCELLATION_DRIVERS = {
    'weather':        'Wind advisory > 15mph forecast',
    'low_engagement': 'Member engagement score declining 3+ weeks',
    'historical':     'Cancelled 2 of last 5 bookings',
    'no_dining':      'No dining visit in 30 days (lower commitment)',
    'complaint_open': 'Open unresolved complaint on file',
}

RECOMMENDED_ACTIONS = {
    'high':   'Personal outreach from Head Pro + lunch reservation offer',
    'medium': 'Automated confirmation nudge with weather update',
    'low':    'Standard confirmation reminder',
}

DAY_OF_WEEK = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']
TIME_BLOCKS = ['7-8 AM', '8-9 AM', '9-10 AM', '10-11 AM',
               '11-12 PM', '12-1 PM', '1-2 PM', '2-3 PM']

# Fill rate patterns by day and time
FILL_RATE_TEMPLATE = {
    'Mon': [0.42, 0.38, 0.35, 0.30, 0.28, 0.25, 0.22, 0.20],
    'Tue': [0.45, 0.40, 0.38, 0.32, 0.30, 0.28, 0.24, 0.20],
    'Wed': [0.50, 0.46, 0.42, 0.38, 0.34, 0.30, 0.26, 0.22],
    'Thu': [0.55, 0.50, 0.46, 0.40, 0.36, 0.32, 0.28, 0.24],
    'Fri': [0.62, 0.58, 0.52, 0.46, 0.40, 0.35, 0.30, 0.26],
    'Sat': [0.96, 0.97, 0.95, 0.88, 0.80, 0.72, 0.62, 0.50],
    'Sun': [0.94, 0.95, 0.93, 0.86, 0.78, 0.68, 0.58, 0.45],
}


def gen_member_waitlist(
    members: list[dict],
    rng: random.Random,
) -> list[tuple]:
    """
    10 waitlist entries: 5 at-risk + 5 healthy contrast members.
    All requesting Saturday January 18, 2026.
    """
    rows = []
    wl_num = 0

    # At-risk members first (retention_priority = HIGH)
    for spec in AT_RISK_MEMBERS:
        wl_num += 1
        days_waiting = rng.randint(3, 7)
        alts = rng.choice(ALT_SLOTS)
        slot = rng.choice(REQUESTED_SLOTS)
        rows.append((
            f'mwl_{wl_num:03d}',
            spec['member_id'],
            'crs_001',
            '2026-01-18',
            slot,
            json.dumps(alts),
            days_waiting,
            'HIGH',
            None,   # notified_at
            None,   # filled_at
            0,      # dining_incentive_attached
        ))

    # Healthy contrast members
    healthy_members = [
        m for m in members
        if m['archetype'] in ('Die-Hard Golfer', 'Balanced Active', 'Social Butterfly',
                               'New Member', 'Snowbird')
        and m['member_id'] not in {s['member_id'] for s in AT_RISK_MEMBERS}
    ]
    sampled = rng.sample(healthy_members, min(5, len(healthy_members)))
    for m in sampled:
        wl_num += 1
        days_waiting = rng.randint(1, 5)
        alts = rng.choice(ALT_SLOTS)
        slot = rng.choice(REQUESTED_SLOTS)
        rows.append((
            f'mwl_{wl_num:03d}',
            m['member_id'],
            'crs_001',
            '2026-01-18',
            slot,
            json.dumps(alts),
            days_waiting,
            'NORMAL',
            None, None, 0,
        ))

    return rows

WAITLIST_COLS = [
    'waitlist_id', 'member_id', 'course_id', 'requested_date', 'requested_slot',
    'alternatives_accepted', 'days_waiting', 'retention_priority',
    'notified_at', 'filled_at', 'dining_incentive_attached',
]


def gen_cancellation_risk(
    booking_rows: list[tuple],
    booking_cols: list[str],
    member_map: dict,
    weather_map: dict,
    rng: random.Random,
) -> list[tuple]:
    """
    Score upcoming confirmed bookings (Jan 17+) for cancellation risk.
    High-risk members: at-risk resignees + bookings on windy days.
    """
    rows = []
    risk_num = 0
    sim_date = date(2026, 1, 17)

    HIGH_RISK_MEMBERS = {'mbr_059', 'mbr_071', 'mbr_072', 'mbr_038', 'mbr_089'}

    for row in booking_rows:
        d = dict(zip(booking_cols, row))
        if d['status'] != 'confirmed':
            continue
        bkg_date = date.fromisoformat(d['booking_date'])
        if bkg_date < sim_date:
            continue

        # Only score bookings that have member players (skip pure-guest groups)
        # We'll score the first member in the group — in practice, cross-reference player_rows
        # but here we use a heuristic based on booking metadata

        # Assign a plausible member_id: use a hash of booking_id to pick deterministically
        mid_seed = int(d['booking_id'].replace('bkg_', '')) % 100
        mid = f'mbr_{(mid_seed + 1):03d}'
        member = member_map.get(mid, {})
        arch = member.get('archetype', 'Balanced Active')

        # Base cancel probability
        if mid in HIGH_RISK_MEMBERS:
            base_prob = rng.uniform(0.55, 0.80)
        elif arch == 'Declining':
            base_prob = rng.uniform(0.30, 0.50)
        elif arch == 'Ghost':
            base_prob = rng.uniform(0.25, 0.45)
        else:
            base_prob = rng.uniform(0.05, 0.20)

        # Weather signal
        wx = weather_map.get(d['booking_date'], {})
        wind_mph = wx.get('wind_mph', 0)
        drivers = []
        if wind_mph > WIND_THRESHOLD_MPH:
            base_prob = min(0.95, base_prob + 0.20)
            drivers.append(CANCELLATION_DRIVERS['weather'])
        if mid in HIGH_RISK_MEMBERS:
            drivers.append(CANCELLATION_DRIVERS['low_engagement'])
        if arch in ('Declining', 'Ghost'):
            drivers.append(CANCELLATION_DRIVERS['historical'])
        if mid == 'mbr_038':
            drivers.append(CANCELLATION_DRIVERS['complaint_open'])
        if not drivers:
            drivers.append(CANCELLATION_DRIVERS['historical'])

        cancel_prob = round(min(0.95, base_prob), 2)

        if cancel_prob >= 0.60:
            action_key = 'high'
        elif cancel_prob >= 0.35:
            action_key = 'medium'
        else:
            action_key = 'low'

        est_revenue_lost = round(rng.uniform(120, 420), 2)

        risk_num += 1
        rows.append((
            f'cr_{risk_num:03d}',
            d['booking_id'],
            mid,
            '2026-01-17T02:00:00',   # scored nightly
            cancel_prob,
            json.dumps(drivers),
            RECOMMENDED_ACTIONS[action_key],
            est_revenue_lost,
            None,   # action_taken
            None,   # outcome
        ))

    return rows

CANCEL_RISK_COLS = [
    'risk_id', 'booking_id', 'member_id', 'scored_at',
    'cancel_probability', 'drivers', 'recommended_action',
    'estimated_revenue_lost', 'action_taken', 'outcome',
]


def gen_demand_heatmap(rng: random.Random) -> list[tuple]:
    """
    56 rows: 2 courses × 7 days × 4 representative time blocks.
    Saturday 7-9 AM is chronically oversubscribed.
    Tue-Thu are underutilized.
    """
    rows = []
    hm_num = 0
    underutil_days = {'Mon', 'Tue', 'Wed', 'Thu'}
    oversubscribed = {('Sat', '7-8 AM'), ('Sat', '8-9 AM'),
                      ('Sun', '7-8 AM'), ('Sun', '8-9 AM')}

    for course_id in ('crs_001', 'crs_002'):
        for dow in DAY_OF_WEEK:
            for block in TIME_BLOCKS:
                hm_num += 1
                key = (dow, block)
                day_idx = DAY_OF_WEEK.index(dow)
                block_idx = TIME_BLOCKS.index(block)

                base_fill = FILL_RATE_TEMPLATE[dow][block_idx]
                noise = rng.uniform(-0.04, 0.04)
                fill = round(max(0.05, min(1.0, base_fill + noise)), 3)

                if key in oversubscribed:
                    fill = round(min(1.0, fill + rng.uniform(0.02, 0.08)), 3)
                    unmet = rng.randint(8, 18)
                    level = 'oversubscribed'
                elif dow in underutil_days and block_idx >= 4:
                    fill = round(max(0.05, fill - 0.10), 3)
                    unmet = 0
                    level = 'underutilized'
                elif fill >= 0.90:
                    unmet = rng.randint(3, 10)
                    level = 'oversubscribed'
                elif fill < 0.45:
                    unmet = 0
                    level = 'underutilized'
                else:
                    unmet = 0
                    level = 'normal'

                rows.append((
                    f'dh_{hm_num:03d}', course_id, dow, block,
                    fill, unmet, level, '2026-01',
                ))

    return rows

HEATMAP_COLS = ['heatmap_id', 'course_id', 'day_of_week', 'time_block',
                 'fill_rate', 'unmet_rounds', 'demand_level', 'computed_for_month']
